src/platform_impl/linux/x11/event_processor.rs

```diff
@@ -615,6 +615,7 @@ impl<T: 'static> EventProcessor<T> {
                 } else {
                     return;
                 };
+                let xkev: &mut ffi::XKeyEvent = xev.as_mut();
```

Comment:
This is an invalid cast. `xev` is an `XGenericEvent` instance in this case.

Reply:
OK, I think this explains the strange behavior I was seeing. Joke's on me for trusting the type system within `x11_dl`.

That said, I'm back to being very confused as to how to track updates to the `device_mod_state` correctly within this event handler.